tests/compiler/test_dict.py

```diff
@@ -428,3 +428,56 @@ def lookup(str):
 
     for var, exp_val in zip(vars_to_lookup, expected_values):
         assert lookup(var) == exp_val
+
+
+@pytest.mark.parametrize(
+    "lookup_path, expectation",
+    [
+        ('a["A"]["1"]', True),
+        ('a["A"]["2"]', False),
+        ('a["A"]["3"]', False),
+        ('a["A"]["4"]', False),
+        ('a["A"]', True),
+        ('a["B"]', False),
+        ('a["C"]', False),
+        ('a["D"]', True),
+        ('a["K"]', False),
+        # ('a["K"]["a"]', False),
+        # ('a["K"]["k"]', False),
+        # ('b', False),
+    ],
+)
+def test_dict_is_defined_4317(snippetcompiler, lookup_path, expectation):
+    """
+    Check that calling "is defined" on dictionaries works as syntactic sugar for != null
+    """
+    snippetcompiler.setup_for_snippet(
+        """
+    entity Test:
+        bool value_is_defined
+    end
+
+    implement Test using std::none
+    t = Test()
+
+    a = {
+        "A": {
+            "1": "ok",
+            "2": null,
+            "3": []
+        },
+        "B": null,
+        "C": [],
+        "D": "ok"
+    }"""
+        f"""if {lookup_path} is defined:
+        t.value_is_defined = true
+    else:
+        t.value_is_defined = false
+    end
+    """
+    )
+    (_, scopes) = compiler.do_compile()
+
+    root = scopes.get_child("__config__")
+    assert root.lookup("t").get_value().lookup("value_is_defined").get_value() == expectation
```

Comment:
This approach is perfectly fine (and is used all over the place), but FYI it might be easier and more compact in the future to just count on a double set exception to catch failures, i.e.
```
is_defined = {str(expectation).lower()}
is_defined = {lookup_path} is defined
```
Then the compilation fails iff the expectation was not met.

Reply:
Oh that's clever :+1:

Reply:
I often name the variable `assert` (or perhaps `assert_expectation` in a case like this), then the purpose becomes really clear: